param_passing/proposal.md

```diff
@@ -0,0 +1,388 @@
+manipulating almost all parameters in SDF models through model/world files
+without requiring modifications to already existing/created model files.
+The new attributes `element_id` and `action` (detailed below) will be reserved
+from being used in other elements in SDFormat.
+
+### 1. Add custom element `//include/experimental:params`
+
+In a world file, the user can construct a model using model composition and
+include components of the model using the `//include` tag. If a user wants to
+modify parameters of the included model, then the user will use the experimental
+custom element `//include/experimental:params` that will describe the
+adjustments to elements. In the custom element `//include/experimental:params`,
+`experimental` is the namespace prefix and `params` is the custom element. This
+approach was chosen so that downstream users can choose to ignore the namespaced
+custom element and its contents. The goal is to implement passing additional
+arguments in `//include` through an experimental custom element in
+SDFormat 1.7 / `libsdformat` 10 then after being vetted it will be made official
+in a future SDFormat / `libsdformat` release where the custom element will be
+changed to `//include/params`.The included model (not the constructed model in
+the model/world file) will be referred to as the original model/file.
+In the model/world file and under `//include`, the elements listed under
+`//include/experimental:params` will indicate elements from the original model
+and will specify new values and/or elements to be updated, added, and/or
+removed.
+
+**Details**
+
+To specify an element from the original file, the tag (e.g., `//link`,
+`//sensor`, `//visual`) needs to be provided as well as the `element_id` attribute
+where the `element_id` is the name of all the parent elements leading to the
+specified element separated by double colons (`::`). For example, given:
+
+```xml
+<link name="chassis">
+  <sensor name="camera">
+    <camera name="cam">
+      ...
+    </camera>
+  </sensor>
+</link>
+```
+
+If the user wants to specify the `//camera` element, then
+`element_id="chassis::camera::cam"` where "chassis" is the name of the `//link`,
+"camera" is the name of the `//link/sensor`, and "cam" is the name of the
+`//link/sensor/camera`. This is how the correct element will be identified and
+will be called the element identifier. Although it is possible to identify the
+element without requiring the user to specify the tag, enforcing the user to
+provide the tag allows less ambiguity to the user as well as provides a means of
+error checking. Then the user will need to specify a corresponding action with
+that element to dictate the alteration they would like to make using a new
+attribute `action`. The attribute takes in a string and the available actions
+are:
+
+* `modify`: indicates that the values in the original model are to be modified
+to the new listed values
```

Comment:
Can you expand on what `modify` does with nested elements?

e.g. if I modify a link's pose, inertia, but I don't replicate its visuals, do they remain?

Reply:
Re-reading, it looks like this may be covered by the examples implicitly (de facto) rather than via an explicitly outlined design (de jure). Is it possible to make ti de jure?

Reply:
How's this? 7c8d6438fab8772a89119f927cdc145cf2948772